src/libs/sys-util-rs/src/numa.rs

```diff
@@ -0,0 +1,210 @@
+    // sysfs directory for CPU devices
+    static ref NUMA_CPU_PATH: String = format!("{}/sys/devices/system/cpu", &*SYS_FS_PREFIX);
+}
+
+const NUMA_NODE_PREFIX: &str = "node";
+const NUMA_NODE_CPU_LIST_NAME: &str = "cpulist";
+
+/// Get numa node id for a CPU
+pub fn get_node_id(cpu: u32) -> Result<u32> {
+    let cpu_path = format!("{}/cpu{}", &*NUMA_CPU_PATH, cpu);
+    let path = Path::new(&cpu_path);
+
+    if let Ok(dirs) = path.read_dir() {
+        for d in dirs.flatten() {
+            if let Ok(file_name) = d.file_name().into_string() {
+                if file_name.starts_with(NUMA_NODE_PREFIX) {
+                    let index_str = file_name.trim_start_matches(NUMA_NODE_PREFIX);
+                    if let Ok(i) = index_str.parse::<u32>() {
+                        return Ok(i);
+                    }
+                }
+            }
+        }
+    } else {
+        return Err(Error::InvalidCpu(cpu))
+    }
+
+    // Default to node 0
+    Ok(0)
+}
+
+/// Map cpulist to NUMA node, returns a HashMap<numa_node_id, Vec<cpu_id>>.
+pub fn get_node_map(cpus: &str) -> Result<HashMap<u32, Vec<u32>>> {
+    // <numa id, Vec<cpu id> >
+    let mut node_map: HashMap<u32, Vec<u32>> = HashMap::new();
+    let cpus_list = split_cpus(cpus)?;
+
+    for c in cpus_list.iter() {
+        let node_id = get_node_id(*c)?;
+        node_map.entry(node_id).or_insert_with(Vec::new).push(*c);
+    }
+
+    Ok(node_map)
+}
+
+/// Get CPU to NUMA node mapping by reading `/sys/devices/system/node/nodex/cpulist`.
+///
+/// Return a HashMap<cpu id, node id>. The hashmap will be empty if NUMA is not enabled on the
+/// system.
+pub fn get_numa_nodes() -> Result<HashMap<u32, u32>> {
+    let mut numa_nodes = HashMap::new();
+    let numa_node_path = std::path::Path::new(&*NUMA_NODE_PATH);
+    if !numa_node_path.exists() {
+        debug!(sl!(), "no numa node available on this system");
+        return Ok(numa_nodes);
+    }
+
+    if let Ok(dirs) = numa_node_path.read_dir() {
+        for d in dirs.flatten() {
+            if let Ok(file_name) = d.file_name().into_string() {
+                if file_name.starts_with(NUMA_NODE_PREFIX) {
+                    let index_string = file_name.trim_start_matches(NUMA_NODE_PREFIX);
+                    info!(
+                        sl!(),
+                        "get node dir {} node index {}", &file_name, index_string
+                    );
+                    match index_string.parse::<u32>() {
+                        Ok(nid) => read_cpu_info_from_node(&d, nid, &mut numa_nodes)?,
+                        Err(_e) => return Err(Error::InvalidNodeFileName(file_name.to_string())),
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(numa_nodes)
+}
+
+fn read_cpu_info_from_node(
+    d: &DirEntry,
+    node_index: u32,
+    numa_nodes: &mut HashMap<u32, u32>,
+) -> Result<()> {
+    let mut cpu_list_path = d.path();
+    cpu_list_path.push(NUMA_NODE_CPU_LIST_NAME);
```

Comment:
`let cpu_list_path = d.path().join(NUMA_NODE_CPU_LIST_NAME)` would probably be simpler.

Reply:
fixed